version.sh

```diff
@@ -1,18 +1,37 @@
 #!/bin/bash
 
 # this versioning algo:
-# keeps tag as is in case if this version is an equal match
-# otherwise adds .<number of commits since last tag>
+#  - if on a tagged commit, use the tag
+#    e.g. 6.2.18 (for the commit tagged 6.2.18)
+#  - if last tag was a regular relase, bump the minor version, make a it a 'dev' pre-release, and append # of commits since tag
+#    e.g. 5.5.38-dev.5+a1b2c3d4 (for commit a1b2c3d4 5 commits after 5.5.37)
+#  - if last tag was a pre-release tag (e.g. alpha, beta, rc), append number of commits since the tag
+#    e.g. 7.0.0-alpha.1.5+a1b2c3d4 (for commit a1b2c3d4 5 commits after 7.0.0-alpha.1)
+
+
+increment_patch() {
+    # increment_patch returns x.y.(z+1) given valid x.y.z semver.
+    # If we need to robustly handle this, it is probably worth
+    # looking at https://github.com/davidaurelio/shell-semver/
+    # or moving this logic to a 'real' programming language -- 2020-03 walt
+    major=$(echo $1 | cut -d'.' -f1)
+    minor=$(echo $1 | cut -d'.' -f2)
+    patch=$(echo $1 | cut -d'.' -f3)
+    patch=$((patch + 1))
+    echo "${major}.${minor}.${patch}"
+}
 
 SHORT_TAG=`git describe --abbrev=0 --tags`
 LONG_TAG=`git describe --tags`
-COMMIT_WITH_LAST_TAG=`git show-ref --tags --dereference | grep "refs/tags/${SHORT_TAG}^{}" | awk '{print $1}'`
+COMMIT_WITH_LAST_TAG=`git rev-list -n1 ${SHORT_TAG}`
 COMMITS_SINCE_LAST_TAG=`git rev-list  ${COMMIT_WITH_LAST_TAG}..HEAD --count`
+SHORT_COMMIT_SHA=`git rev-parse --short HEAD`
 
-if [[ "$LONG_TAG" == "$SHORT_TAG" ]] ; then
+if [[ "$LONG_TAG" == "$SHORT_TAG" ]] ; then  # the current commit is tagged as a release
     echo "$SHORT_TAG"
-elif [[ "$SHORT_TAG" != *-* ]] ; then
-    echo "$SHORT_TAG-${COMMITS_SINCE_LAST_TAG}"
-else
-    echo "$SHORT_TAG.${COMMITS_SINCE_LAST_TAG}"
+elif [[ "$SHORT_TAG" != *-* ]] ; then  # the current ref is not a decendent of a pre-release version
+    SHORT_TAG=$(increment_patch ${SHORT_TAG})
+    echo "$SHORT_TAG-dev.${COMMITS_SINCE_LAST_TAG}+${SHORT_COMMIT_SHA}"
```

Comment:
So I'm on a fence with that change: for one this is definitely a better differentiator since it will allow to tell between two branches equally away from a tag, on the other - as you pointed out whether this is compatible with all other use cases of a version.
I'd be fine with it as long as it worked in all scenarios.

Reply:
>  I'd be fine with it as long as it worked in all scenarios.

I'll put some more effort into it this morning, and see if I can tweak our make files in a reasonable way to work with docker (and any other scenarios where `+metatdata` breaks).  I'd prefer to avoid `_` in the canonical build identifier, as that will mean the build is no longer semver compliant

Reply:
I haven't found a quick fix yet.  As such, I've filed #1192 to track this long term work, and removed 4bf85f07bdc2a336ca91209b38d9c6a26f9ecefc from this PR, such that we can move forward.